src/CodeGen_PTX_Dev.cpp

```diff
@@ -385,157 +388,135 @@ void CodeGen_PTX_Dev::visit(const Atomic *op) {
 
 void CodeGen_PTX_Dev::codegen_vector_reduce(const VectorReduce *op, const Expr &init) {
     // Pattern match 8/16-bit dot products
+    struct Pattern {
+        VectorReduce::Operator op;
+        int factor;
+        Expr pattern;
+        const char *name;
+    };
+    static Expr wild_i8x = Variable::make(Int(8, 0), "*");
+    static Expr wild_u8x = Variable::make(UInt(8, 0), "*");
+    static Expr wild_i16x = Variable::make(Int(16, 0), "*");
+    static Expr wild_u16x = Variable::make(UInt(16, 0), "*");
+    // clang-format off
+    static const Pattern patterns[] = {
+        {VectorReduce::Add, 4, i32(widening_mul(wild_i8x, wild_i8x)), "dp4a_s32_s32"},
+        {VectorReduce::Add, 4, i32(widening_mul(wild_i8x, wild_u8x)), "dp4a_s32_u32"},
+        {VectorReduce::Add, 4, i32(widening_mul(wild_u8x, wild_i8x)), "dp4a_u32_s32"},
+        {VectorReduce::Add, 4, u32(widening_mul(wild_u8x, wild_u8x)), "dp4a_u32_u32"},
+        {VectorReduce::Add, 2, widening_mul(wild_i16x, wild_i16x), "dp2a_s32_s32"},
+        {VectorReduce::Add, 2, widening_mul(wild_i16x, wild_u16x), "dp2a_s32_u32"},
+        {VectorReduce::Add, 2, widening_mul(wild_u16x, wild_i16x), "dp2a_u32_s32"},
+        {VectorReduce::Add, 2, widening_mul(wild_u16x, wild_u16x), "dp2a_u32_u32"},
+    };
+    // clang-format on
 
     const int input_lanes = op->value.type().lanes();
     const int factor = input_lanes / op->type.lanes();
-    const Mul *mul = op->value.as<Mul>();
-    if (op->op == VectorReduce::Add &&
-        mul &&
-        (factor % 4 == 0) &&
-        (op->type.element_of() == Int(32) ||
-         op->type.element_of() == UInt(32))) {
-        Expr i = init;
-        if (!i.defined()) {
-            i = cast(mul->type, 0);
+
+    std::vector<Expr> matches;
+    for (const Pattern &p : patterns) {
+        if (p.op != op->op || factor % p.factor != 0) {
+            continue;
         }
-        // Try to narrow the multiply args to 8-bit
-        Expr a = mul->a, b = mul->b;
-        if (op->type.is_uint()) {
-            a = lossless_cast(UInt(8, input_lanes), a);
-            b = lossless_cast(UInt(8, input_lanes), b);
-        } else {
-            a = lossless_cast(Int(8, input_lanes), a);
-            b = lossless_cast(Int(8, input_lanes), b);
-            if (!a.defined()) {
-                // try uint
-                a = lossless_cast(UInt(8, input_lanes), mul->a);
-            }
-            if (!b.defined()) {
-                b = lossless_cast(UInt(8, input_lanes), mul->b);
-            }
+        if (!expr_match(p.pattern, op->value, matches)) {
```

Comment:
What stopped you from using IRMatch patterns here?

Reply:
I don't think we can rewrite to calls that aren't intrinsics?

Reply:
Ah. Please add a TODO then.

Reply:
(to add that feature to IRMatch)